src/main/java/com/hierynomus/ntlm/functions/NtlmFunctions.java

```diff
@@ -148,6 +155,77 @@
         }
     }
 
+    /**
+     * [MS-NLMP].pdf 2.2.2.7 NTLM v2: NTLMv2_CLIENT_CHALLENGE
+     *
+     * 3.3.2 NTLM v2 Authentication
+     * Set temp to ConcatenationOf(Responserversion, HiResponserversion, Z(6), Time, ClientChallenge, Z(4), ServerName, Z(4))
+     *
+     * @param challengeFromClient
+     * @param targetInformation
+     * @return
+     */
+    public static byte[] getNTLMv2ClientChallenge(
+                                                  byte[] challengeFromClient,
+                                                  byte[] targetInformation
+                                                  ) {
+        if (challengeFromClient == null) {
+            return null;
+        }
+        long nanos1601 = getTimeStampNanos1601();
+        byte[] l_targetInfo = (targetInformation == null) ? new byte[0] : targetInformation;
+        Buffer.PlainBuffer ccBuf = new Buffer.PlainBuffer(Endian.LE);
+        ccBuf.putByte((byte)0x01); // RespType (1)
+        ccBuf.putByte((byte)0x01); // HiRespType (1)
+        ccBuf.putUInt16(0); // Reserved1 (2)
+        ccBuf.putUInt32(0); // Reserved2 (4)
+        ccBuf.putLong(nanos1601); // Timestamp (8)
+        ccBuf.putRawBytes(challengeFromClient); // ChallengeFromClient (8)
+        ccBuf.putUInt32(0); // Reserver3 (4)
+        ccBuf.putRawBytes(l_targetInfo);
+        ccBuf.putUInt32(0); // Last AV Pair indicator
+
+        return ccBuf.getCompactData();
+    }
+
+    /**
+     *
+     * 3.3.2 NTLM v2 Authentication
+     *
+     * Set NTProofStr to HMAC_MD5(ResponseKeyNT, ConcatenationOf(CHALLENGE_MESSAGE.ServerChallenge,temp))
+     * Set NtChallengeResponse to ConcatenationOf(NTProofStr, temp)
+     *
+     * @param responseKeyNT
+     * @param serverChallenge
+     * @param ntlmv2ClientChallenge (temp from above)
+     * @return
+     */
+    public static byte[] getNTLMv2Response(byte[] responseKeyNT, byte[] serverChallenge, byte[] ntlmv2ClientChallenge) {
+
+        byte[] ntProofStr = hmac_md5(responseKeyNT, serverChallenge, ntlmv2ClientChallenge);
+
+        byte[] ntChallengeResponse = ByteBuffer.allocate(ntProofStr.length + ntlmv2ClientChallenge.length)
+                .put(ntProofStr).put(ntlmv2ClientChallenge).array();
+
+        return ntChallengeResponse;
+    }
+
+    public static byte[] encryptRc4(byte[] key, byte[] val) throws BadPaddingException, IllegalBlockSizeException {
+        Cipher c = getRC4Cipher(key);
+        byte[] enc = c.doFinal(val);
+        return enc;
+    }
+
+    /**
+     * A 64-bit unsigned integer that contains the current system time, represented
+     * as the number of 100 nanosecond ticks elapsed since midnight of January 1, 1601 (UTC)
+     */
+    private static long getTimeStampNanos1601() {
+        final long MILLISECONDS_BETWEEN_1970_AND_1601 = 11644473600000L;
```

Comment:
`MsDataTypes.java` has this info. Maybe make a method there called `nowAsFileTime` and use that (the official denominator for this timestamp is "Windows FILETIME" I think)